cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+*/
+import "C"
+
+import (
+	"context"
+	"errors"
+	"time"
+	"unsafe"
+)
+
+// CartoLib holds the c type viam_carto_lib
+type CartoLib struct {
+	value *C.viam_carto_lib
+}
+
+// Carto holds the c type viam_carto
+type Carto struct {
+	value *C.viam_carto
+}
+
+// AddSensorReading holds the c type viam_carto_sensor_reading
+type AddSensorReading struct {
+	value C.viam_carto_sensor_reading
+}
+
+// GetPositionResponse holds the c type viam_carto_get_position_response
+type GetPositionResponse struct {
+	value C.viam_carto_get_position_response
+}
+
+// GetPositionHolder holds values returned from c to be processed later
+type GetPositionHolder struct {
+	x float64
+	y float64
+	z float64
+
+	ox    float64
+	oy    float64
+	oz    float64
+	theta float64
+
+	real float64
+	imag float64
+	jmag float64
+	kmag float64
+
+	compReference string
+}
+
+// GetInternalStateResponse holds the c type viam_carto_get_internal_state_response
+type GetInternalStateResponse struct {
+	value C.viam_carto_get_internal_state_response
+}
+
+// LidarConfig represents the lidar configuration
+type LidarConfig int64
+
+const (
+	twoD LidarConfig = iota
+	threeD
+)
+
+// CartoConfig contains config values from app
+type CartoConfig struct {
+	sensors            []string
+	mapRateSecond      int
+	dataDir            string
+	componentReference string
+	lidarConfig        LidarConfig
+}
+
+// CartoAlgoConfig contains config values from app
+type CartoAlgoConfig struct {
+	optimizeOnStart      bool
+	optimizeEveryNNodes  int
+	numRangeData         int
+	missingDataRayLength float64
+	maxRange             float64
+	minRange             float64
+	maxSubmapsToKeep     int
+	freshSubmapsCount    int
+	minCoveredArea       float64
+	minAddedSubmapsCount int
+	occupiedSpaceWeight  float64
+	translationWeight    float64
+	rotationWeight       float64
+}
+
+// NewCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewCartoLib(miniloglevel, verbose int) (CartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(&pVcl))
+	/*
+		Question: what is going on here? are we not getting correct amount of indirection?
+					fmt.Printf("%#v\n", pVcl) = &cartofacade._Ctype_struct_viam_carto_lib{minloglevel:0, verbose:0}
+					fmt.Println(&pVcl) = 0x14000128078
+					fmt.Println(goPtr) = 0x600001ee8000
+					// ppVcl := (**C.viam_carto_lib)(goPtr)
+					fmt.Println(ppVcl) = 0x600001ee8000
+	*/
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return CartoLib{}, err
+	}
+
+	vcl := CartoLib{value: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CartoLib) TerminateCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
```

Comment:
I don't understand how this can work. `vcl` is a `CartoLib{}` which is not a valid input into `viam_carto_lib_terminate`.

Reply:
this was definitely a mistake, I updated this line to be:

`status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(&vcl.value)))`

I suspect casting the untyped memory is why it worked, but it definitely would have been incorrect and think that we should ask eric about this as well